### PR TITLE
[TT-2539] added access/transaction logs

### PR DESCRIPTION
**Pull request by LLe27**
### **FR Jira Ticket**

https://tyktech.atlassian.net/browse/TT-2539

## Description

- Added the `TYK_GW_ACCESSLOGS_ENABLED` Gateway config option
- The Tyk Gateway will determine to print access logs to STDOUT for both success and error handling situations
  - If the `TYK_GW_ACCESSLOGS_ENABLED` is set to `true` then the Gateway will print access logs to STDOUT
  - If the `TYK_GW_ACCESSLOGS_ENABLED` is set to `false` then the Gateway will not print access logs to STDOUT

Note that this feature is off by default and that the `AccessLog` struct only contains the more common elements. Below are some examples of an access log

```
time="Sep 04 08:04:18" level=info APIID=c062396cb62d4e9a5ee37adaf85b9e4c APIKey=00000000 ClientIP=127.0.0.1 ClientRemoteAddr="127.0.0.1:53506" Host="localhost:8080" Method=GET OrgID=66d07f00247d80811d5199c3 Proto=HTTP/1.1 RequestURI=/httpbin/get StatusCode=200 TotalLatency=381 UpstreamAddress="http://httpbin.org/get" UpstreamLatency=381 UpstreamPath=/get UpstreamURI=/get UserAgent=curl/8.1.2 prefix=access-log
time="Sep 04 08:08:20" level=info APIID=c062396cb62d4e9a5ee37adaf85b9e4c APIKey=00000000 ClientIP=127.0.0.1 ClientRemoteAddr="127.0.0.1:53566" Host="localhost:8080" Method=GET OrgID=66d07f00247d80811d5199c3 Proto=HTTP/1.1 RequestURI=/httpbin/get StatusCode=401 TotalLatency=0 UpstreamAddress=":///httpbin/get" UpstreamLatency=0 UpstreamPath=/httpbin/get UpstreamURI=/httpbin/get UserAgent=curl/8.1.2 prefix=access-log
```

## Motivation and Context

Today the Tyk Gateway does not print access logs for success API calls but instead only for error API calls. Providing access logs for both scenarios within the Tyk Gateway is extremely valuable especially if you are monitoring logs, capturing analytics or even debugging. Providing the option to turn on or off the Tyk Gateway access logs will provide clients more insights in for API calls in regards to success and error situations.

## How This Has Been Tested

- Manual testing
- Unit testing
- Performance testing/benchmarks

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

### CODE REVIEW

File: gateway/handler_success_test.go [hunk 1]
```diff
@@ -322,3 +322,57 @@ func TestAnalyticsIgnoreSubgraph(t *testing.T) {
 	)
 	assert.NoError(t, err)
 }
+
+func BenchmarkSuccessLogTransaction(b *testing.B) {
```
Comment (LLe27) [inline gateway/handler_success_test.go:326]: Current benchmark with `-benchtime=10s`

```
BenchmarkSuccessLogTransaction/AccessLogs_enabled_with_Hashkeys_set_to_true-10         	  101232	    111830 ns/op	   35012 B/op	     338 allocs/op
BenchmarkSuccessLogTransaction/AccessLogs_enabled_with_Hashkeys_set_to_false
BenchmarkSuccessLogTransaction/AccessLogs_enabled_with_Hashkeys_set_to_false-10        	  101864	    112778 ns/op	   34822 B/op	     329 allocs/op
BenchmarkSuccessLogTransaction/AccessLogs_disabled_with_Hashkeys_set_to_true
BenchmarkSuccessLogTransaction/AccessLogs_disabled_with_Hashkeys_set_to_true-10        	  108433	    108124 ns/op	   32122 B/op	     315 allocs/op
BenchmarkSuccessLogTransaction/AccessLogs_disabled_with_Hashkeys_set_to_false
BenchmarkSuccessLogTransaction/AccessLogs_disabled_with_Hashkeys_set_to_false-10       	  108339	    110007 ns/op	   32032 B/op	     310 allocs/op
```

File: gateway/handler_error_test.go [hunk 2]
```diff
@@ -122,3 +123,50 @@ func TestHandleDefaultErrorJSON(t *testing.T) {
 	})
 
 }
+
+func BenchmarkErrorLogTransaction(b *testing.B) {
```
Comment (LLe27) [inline gateway/handler_error_test.go:127]: Current benchmark with `-benchtime=10s`

```
BenchmarkErrorLogTransaction/AccessLogs_enabled_with_Hashkeys_set_to_true-10         	  268650	     38135 ns/op	    7979 B/op	      84 allocs/op
BenchmarkErrorLogTransaction/AccessLogs_enabled_with_Hashkeys_set_to_false
BenchmarkErrorLogTransaction/AccessLogs_enabled_with_Hashkeys_set_to_false-10        	  300806	     36877 ns/op	    7970 B/op	      84 allocs/op
BenchmarkErrorLogTransaction/AccessLogs_disabled_with_Hashkeys_set_to_true
BenchmarkErrorLogTransaction/AccessLogs_disabled_with_Hashkeys_set_to_true-10        	  308988	     37358 ns/op	    7969 B/op	      84 allocs/op
BenchmarkErrorLogTransaction/AccessLogs_disabled_with_Hashkeys_set_to_false
BenchmarkErrorLogTransaction/AccessLogs_disabled_with_Hashkeys_set_to_false-10       	  298038	     37823 ns/op	    7972 B/op	      84 allocs/op
```

File: internal/httputil/access_log.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+func NewAccessLogRecord(apiID string, apiKey string, orgId string) *AccessLogRecord {
+	return &AccessLogRecord{
+		APIID:  apiID,
+		APIKey: apiKey,
+		OrgID:  orgId,
+	}
+}
+
+// Logger renders a single access log entry
+func (a *AccessLogRecord) Logger(log *logrus.Logger) *logrus.Entry {
```
Comment (jeffy-mathew) [inline internal/httputil/access_log.go:32]: Hi, 
can we make this `Fields` method and return `logrus.Fields` instead of `*Logrus.Entry`? So that `log *logrus.Logger` need not be passed.
the caller would look like 
```
log.WithFields(accessLog.Fields()).Info()
```

File: gateway/handler_error.go [hunk 4]
```diff
@@ -311,6 +314,29 @@ func (e *ErrorHandler) HandleError(w http.ResponseWriter, r *http.Request, errMs
...
+		} else {
+			key = storage.HashKey(token, hashKeys)
+		}
+
+		accessLog := httputil.NewAccessLogRecord(e.Spec.APIID, key, e.Spec.OrgID)
+		accessLog.WithLatency(&analytics.Latency{})
+		accessLog.WithRequest(r)
+		accessLog.WithResponse(response)
+
+		logger := accessLog.Logger(log)
```
Comment (titpetric) [inline gateway/handler_error.go:336]: It would be better to return logrus.Fields from this;

Expected: `log.WithFields(accessLog.Fields()).Info()`

- Rename `accessLog.Logger(*logrus.Logger) *logrus.Entry` into `Fields() logrus.Fields`

File: gateway/handler_error.go [hunk 5]
```diff
@@ -311,6 +314,29 @@ func (e *ErrorHandler) HandleError(w http.ResponseWriter, r *http.Request, errMs
 			log.WithError(err).Error("could not store analytic record")
 		}
 	}
+
```
Comment (titpetric) [inline gateway/handler_error.go:317]: Expecting one line added here: `e.recordAccessLog(r.Context(), log)`, move all this code into own function. The logger is passed along because otherwise it's a global coupling. The handlers don't yet have their own logger (or one passed down).

Logic here slightly differs, ensure consistent implementation between types, as I understand the logic for both should be identical (invoking ctxGetAuthToken based on ctx).

File: gateway/handler_error.go [hunk 6]
```diff
@@ -311,6 +314,29 @@ func (e *ErrorHandler) HandleError(w http.ResponseWriter, r *http.Request, errMs
...
+	if e.Spec.GlobalConfig.AccessLogs.Enabled {
+		// Don't print the full token, handle as obfuscated key or hashed key for security reasons
+		hashKeys := e.Gw.GetConfig().HashKeys
+		key := ""
+
+		if !hashKeys {
+			key = e.Gw.obfuscateKey(token)
+		} else {
+			key = storage.HashKey(token, hashKeys)
+		}
```
Comment (titpetric) [inline gateway/handler_error.go:329]: This block just seems like utter spaghetti code. @sredxny why wouldn't obfuscateKey handle this?

https://github.com/TykTechnologies/tyk/blob/fd1a199bbf3b99796e13a0444db3bd5d0ec29e8f/gateway/log_helpers.go#L14

File: gateway/handler_success_test.go [hunk 7]
```diff
@@ -322,3 +322,57 @@ func TestAnalyticsIgnoreSubgraph(t *testing.T) {
...
+		conf := func(globalConf *config.Config) {
+			globalConf.HashKeys = false
+			globalConf.AccessLogs.Enabled = false
+		}
+		benchmarkSuccessLogTransaction(b, conf)
+	})
+}
+
+func benchmarkSuccessLogTransaction(b *testing.B, conf func(globalConf *config.Config)) {
+	b.ReportAllocs()
```
Comment (titpetric) [inline gateway/handler_success_test.go:359]: missing b.Helper()

File: internal/httputil/access_log.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/TykTechnologies/tyk-pump/analytics"
+	"github.com/TykTechnologies/tyk/request"
+)
+
+// AccessLogRecord is a representation of a transaction log in the Gateway
+type AccessLogRecord struct {
+	APIID    string
```
Comment (titpetric) [inline internal/httputil/access_log.go:14]: Delete all the fields and just have a `Fields logrus.Fields` value;

File: internal/httputil/access_log.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+// NewAccessLogRecord returns an AccessLogRecord object
+func NewAccessLogRecord(apiID string, apiKey string, orgId string) *AccessLogRecord {
+	return &AccessLogRecord{
+		APIID:  apiID,
+		APIKey: apiKey,
+		OrgID:  orgId,
+	}
+}
```
Comment (titpetric) [inline internal/httputil/access_log.go:29]:
```suggestion
func NewAccessLogRecord(apiID string, apiKey string, orgId string) *AccessLogRecord {
	r := &AccessLogRecord{}
	r.Fields =  logrus.Fields{
		"APIID":  a.APIID,
		"APIKey": a.APIKey,
		"OrgID":  a.OrgID,
		"prefix": "access-log",
	}
	return r
}
```

I think the keys still need renaming (see other feedback)

File: internal/httputil/access_log.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+		fields["UpstreamAddress"] = a.Request.URL.Scheme + "://" + a.Request.URL.Host + a.Request.URL.RequestURI()
+		fields["UpstreamPath"] = a.Request.URL.Path
+		fields["UpstreamURI"] = a.Request.URL.RequestURI()
+		fields["UserAgent"] = a.Request.UserAgent()
+	}
+
+	// Include response field if it exists
+	if a.Response != nil {
+		fields["StatusCode"] = a.Response.StatusCode
+	}
```
Comment (titpetric) [inline internal/httputil/access_log.go:64]: move to with* functions below, populating `a.Fields`.
delete Logger function completely and use the a.Fields from gateway scope

@jeffy-mathew i don't think there's a reason to protect this, it's a throw away allocation. We could have a more logrus.Fields oriented structure and possibly drop AccessLogRecord altogether? `type AccessLogRecord logrus.Fields` directly perhaps?

Comment (jeffy-mathew) [reply internal/httputil/access_log.go:64]: yes, that makes sense

**Comment by andrei-tyk** [comment]
@LLe27  sorry for the huge delay but we've been lacking manpower in the past two weeks due to holidays and other issues. Can you please resolve the conflicts and message me on Slack when it's done so I can leave an approval if everything works/looks right?

**Comment by LLe27** [comment]
@andrei-tyk thanks I've fixed the merge requests -- please take a look and let me know if I can further assist. Please provide any feedback if any! Appreciate it.

File: .gitignore [hunk 11]
```diff
@@ -50,6 +50,7 @@ session_state_gen_test.go
 __pycache__/
 tyk.test
 tyk-gateway.pid
+*.go-e
```
Comment (jeffy-mathew) [inline .gitignore:53]: what is this gitignore for?

File: gateway/handler_error.go [hunk 12]
```diff
@@ -87,6 +92,25 @@ type TemplateExecutor interface {
 	Execute(wr io.Writer, data interface{}) error
 }
 
+func (e *ErrorHandler) recordAccessLog(log *logrus.Logger, req *http.Request, resp *http.Response) {
```
Comment (jeffy-mathew) [inline gateway/handler_error.go:95]: We could avoid repeating `recordAccessLog` on `ErrorHandler` and `SuccessHandler` by refactoring `recordAccessLog` to be a receiver method on `BaseMiddleware` which is embedded in `SuccessHandler` and `ErrorHandler`. 
```
func (t *BaseMiddleware) recordAccessLog(latency *analytics.Latency, log *logrus.Logger, req *http.Request, resp *http.Response) {
	// Don't print the full token, handle as obfuscated key or hashed key for security reasons
	hashKeys := t.Gw.GetConfig().HashKeys
	token := ctxGetAuthToken(req)

	if !hashKeys {
		token = t.Gw.obfuscateKey(token)
	} else {
		token = storage.HashKey(token, hashKeys)
	}

	accessLog := httputil.NewAccessLogRecord(t.Spec.APIID, token, t.Spec.OrgID)
	accessLog.WithLatency(latency)
	accessLog.WithRequest(req)
	accessLog.WithResponse(resp)

	log.WithFields(accessLog.Fields()).Info()
}
```

File: .gitignore (hunk 11 above)
Comment (titpetric) [reply .gitignore:53]: some IDE or mac specific linter artefact, tried sourcing the extension via sourcegraph, didn't find it; files added into git by mistake then removed during review

File: gateway/handler_error.go (hunk 12 above)
Comment (titpetric) [reply gateway/handler_error.go]: fixed some api signature smells, and added an WithAuthToken to access log record;

File: config/config.go [hunk 13]
```diff
@@ -518,10 +524,7 @@ type CoProcessConfig struct {
 
 type CertificatesConfig struct {
 	API []string `json:"apis"`
-	// Upstream is used to specify the certificates to be used in mutual TLS connections to upstream services. These are set at gateway level as a map of domain -> certificate id or path.
-	// For example if you want Tyk to use the certificate `ab23ef123` for requests to the `example.com` upstream and `/certs/default.pem` for all other upstreams then:
-	// In `tyk.conf` you would configure `"security": {"certificates": {"upstream": {"*": "/certs/default.pem", "example.com": "ab23ef123"}}}`
-	// And if using environment variables you would set this to `*:/certs/default.pem,example.com:ab23ef123`.
+	// Specify upstream mutual TLS certificates at a global level in the following format: `{ "<host>": "<cert>" }``
```
Comment (jeffy-mathew) [inline config/config.go:527]: this change is unrelated, right?

**Review by titpetric — APPROVED** [review]

**Comment by LLe27** [comment]
Re-opening merge PR
